### PR TITLE
Changing baked data to use zerotrie

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

**Comment by sffc** [comment]
Have you measured impact on binary size and performance?

**Comment by sffc** [comment]
While I think we should explore this, there are some factors at play that might make ZeroTrie less compelling for baked data than for blob data:

1. The compiler is able to deduplicate locale strings across DataProvider lookup tables, which it did not do in blob1
2. `binary_search` is a primitive that the compiler could figure out how to optimize on its own, whereas ZeroTrie provides only one algorithm for the lookup

My suspicion is that despite these factors, ZeroTrie will still be faster and smaller overall, although to a lesser degree than it was in blob1 vs blob2. I would however like to see some figures.

**Review by sffc — COMMENTED** [review]
^ Pending metrics.

**Comment by sffc** [comment]
Do you have metrics?

**Comment by robertbastian** [comment]
Yes in the diffs

**Comment by sffc** [comment]
I'm sorry, can you help me know which commits to review and which commits are datagen only?

If I omit just the last commit entitled "gen", I still get a really big diff.

image

**Comment by robertbastian** [comment]
You should review #5161 first, as it says in the description.

File: provider/baked/tests/data/hello_world_v1_marker.rs.data [hunk 1]
```diff
@@ -14,37 +14,39 @@ macro_rules! __impl_hello_world_v1_marker {
...
+            const DATA_HELLO_WORLD_V1_MARKER: icu_provider_baked::zerotrie::Data<icu_provider::hello_world::HelloWorldV1Marker> = icu_provider_baked::zerotrie::Data {
+                trie: icu_provider_baked::zerotrie::ZeroTrieSimpleAscii::from_store(&[205u8, 98u8, 99u8, 100u8, 101u8, 102u8, 105u8, 106u8, 108u8, 112u8, 114u8, 115u8, 118u8, 122u8, 2u8, 4u8, 10u8, 66u8, 73u8, 76u8, 88u8, 91u8, 94u8, 102u8, 112u8, 115u8, 110u8, 130u8, 115u8, 131u8, 101u8, 132u8, 45u8, 65u8, 84u8, 133u8, 195u8, 108u8, 110u8, 111u8, 1u8, 49u8, 134u8, 135u8, 194u8, 30u8, 45u8, 8u8, 114u8, 101u8, 118u8, 101u8, 114u8, 115u8, 101u8, 128u8, 195u8, 48u8, 49u8, 71u8, 12u8, 15u8, 194u8, 48u8, 49u8, 6u8, 194u8, 49u8, 50u8, 1u8, 136u8, 137u8, 57u8, 138u8, 52u8, 50u8, 139u8, 66u8, 140u8, 45u8, 117u8, 45u8, 115u8, 100u8, 45u8, 103u8, 98u8, 101u8, 110u8, 103u8, 141u8, 142u8, 194u8, 97u8, 105u8, 1u8, 143u8, 144u8, 0u8, 115u8, 144u8, 1u8, 97u8, 144u8, 2u8, 30u8, 114u8, 101u8, 118u8, 101u8, 114u8, 115u8, 101u8, 129u8, 97u8, 144u8, 3u8, 116u8, 144u8, 4u8, 194u8, 111u8, 117u8, 2u8, 144u8, 5u8, 144u8, 6u8, 114u8, 144u8, 7u8, 45u8, 76u8, 97u8, 116u8, 110u8, 144u8, 8u8, 105u8, 144u8, 9u8, 104u8, 144u8, 10u8]),
```
Comment (sffc) [inline provider/baked/tests/data/hello_world_v1_marker.rs.data:18]: Thought: Maybe out of scope for this PR, but it's a lot of wasted space to repeat `u8` on every item. You only need it once.

Suggestion (optional): the bytes that are ASCII could be represented as `b'x'`

File: utils/zerotrie/src/reader.rs [hunk 2]
```diff
@@ -619,7 +619,8 @@ use alloc::vec::Vec;
 
 /// Internal iterator type for walking the strings contained in a ZeroTrie.
 #[cfg(feature = "alloc")]
-pub(crate) struct ZeroTrieIterator<'a> {
+#[derive(Debug)]
+pub struct ZeroTrieIterator<'a> {
```
Comment (sffc) [inline utils/zerotrie/src/reader.rs:623]: Nit: Update docs to say it's no longer internal.

File: utils/zerotrie/src/zerotrie.rs [hunk 3]
```diff
@@ -628,24 +629,28 @@ impl_zerotrie_subtype!(
     ZeroTrieSimpleAscii,
     String,
     reader::get_iter_ascii_or_panic,
+    core::iter::Map<reader::ZeroTrieIterator<'_>, fn((Vec<u8>, usize)) -> (String, usize)>,
```
Comment (sffc) [inline utils/zerotrie/src/zerotrie.rs:632]: Nit: I'm in support of changing the `impl Iterator` to a concrete type, but I don't want to restrict the concrete type to being a specific map composition like this. Please introduce a new type such as `ZeroTrieAsciiIterator` or something like that so we can change the internals in the future. Safest would be to have a different type for all four concrete impls.

File: provider/baked/tests/data/hello_world_v1_marker.rs.data (hunk 1 above)
Comment (robertbastian) [reply provider/baked/tests/data/hello_world_v1_marker.rs.data:18]: Ah yes I should use `impl Bake for &[u8]`, which uses byte strings.

File: utils/zerotrie/src/zerotrie.rs (hunk 3 above)
Comment (robertbastian) [reply utils/zerotrie/src/zerotrie.rs:632]: And these four types redirect all 76 iterator methods to the types they wrap?

File: utils/databake/src/primitives.rs [hunk 4]
```diff
@@ -121,15 +121,14 @@ fn byte_string(bytes: &[u8]) -> TokenStream {
     // escape slash when trying to match the first "\0" with another digit, and then seeing b"\01".
```
Comment (robertbastian) [inline utils/databake/src/primitives.rs:121]: I'm inclined to remove this, as our MSRV is not past 1.70, and this has apparently been generating iffy code (attribute on expression), with nobody complaining. wdyt?

Comment (sffc) [reply utils/databake/src/primitives.rs:121]: Whatever makes the nicest output and fewest warnings 🤷

**Review by sffc — DISMISSED** [review]
Praise: Impressive size reductions, and this is a fair comparison when converting TinyStr to ZeroTrie since those are both reference-less structures.

File: utils/zerotrie/src/zerotrie.rs (hunk 3 above)
Comment (sffc) [reply utils/zerotrie/src/zerotrie.rs:632]: Fair point, we can clean this up later before zerotrie 1.0

File: utils/zerotrie/src/reader.rs (hunk 2 above)
Comment (sffc) [reply utils/zerotrie/src/reader.rs:623]: ^

File: utils/zerotrie/src/zerotrie.rs (hunk 3 above)
Comment (robertbastian) [reply utils/zerotrie/src/zerotrie.rs]: I've made a `#[doc(hidden)]` type alias, should be non-clickable in the docs

**Review by sffc — APPROVED** [review]
Hash map instability was creeping in somewhere?

**Pull request merged**